src/mgr/QueryParams.cc

```diff
@@ -66,60 +69,59 @@ Mgr::QueryParams::find(const String& name) const
 }
 
 bool
-Mgr::QueryParams::ParseParam(const String& paramStr, Param& param)
+Mgr::QueryParams::ParseParam(SBuf &paramStr, Param &param)
 {
-    bool parsed = false;
-    regmatch_t pmatch[3];
-    regex_t intExpr;
-    regcomp(&intExpr, "^([a-z][a-z0-9_]*)=([0-9]+((,[0-9]+))*)$", REG_EXTENDED | REG_ICASE);
-    regex_t stringExpr;
-    regcomp(&stringExpr, "^([a-z][a-z0-9_]*)=([^&= ]+)$", REG_EXTENDED | REG_ICASE);
-    if (regexec(&intExpr, paramStr.termedBuf(), 3, pmatch, 0) == 0) {
-        param.first = paramStr.substr(pmatch[1].rm_so, pmatch[1].rm_eo);
-        std::vector<int> array;
-        int n = pmatch[2].rm_so;
-        for (int i = n; i < pmatch[2].rm_eo; ++i) {
-            if (paramStr[i] == ',') {
-                array.push_back(atoi(paramStr.substr(n, i).termedBuf()));
-                n = i + 1;
-            }
-        }
-        if (n < pmatch[2].rm_eo)
-            array.push_back(atoi(paramStr.substr(n, pmatch[2].rm_eo).termedBuf()));
-        param.second = new IntParam(array);
-        parsed = true;
-    } else if (regexec(&stringExpr, paramStr.termedBuf(), 3, pmatch, 0) == 0) {
-        param.first = paramStr.substr(pmatch[1].rm_so, pmatch[1].rm_eo);
-        param.second = new StringParam(paramStr.substr(pmatch[2].rm_so, pmatch[2].rm_eo));
-        parsed = true;
+    static const CharacterSet nameChars = CharacterSet("param-name", "_") + CharacterSet::ALPHA + CharacterSet::DIGIT;
+
+    Parser::Tokenizer tok(paramStr);
+
+    SBuf name;
+    if (!tok.prefix(name, nameChars))
+        return false;
+
+    if (!tok.skip('='))
+        return false; // TODO: support name-only parameters
+
+    static const CharacterSet badValueChars = CharacterSet("param-value-bad", "&= ");
+    SBuf valueStr = tok.remaining();
+    if (valueStr.findFirstOf(badValueChars) != SBuf::npos)
+        return false;
+
+    param.first = SBufToString(name);
+
+    std::vector<int> array;
+    int64_t intVal = 0;
+    while (tok.int64(intVal, 10, false)) {
+        static const CharacterSet comma("comma",",");
+        (void)tok.skipAll(comma);
+        Must(intVal >= std::numeric_limits<int>::min());
+        Must(intVal <= std::numeric_limits<int>::max());
+        array.emplace_back(int(intVal));
```

Comment:
The code is presenting constructor parameter - which is an int. Thus the int() conversion.
The first suggestion is not correct because the constructor parameter for array<int> is not array<int64_t>.

Reply:
> The first suggestion is not correct because the constructor parameter for array<int> is not array<int64_t>.

I believe there is a misunderstanding of the problem I am trying to highlight and of what emplace methods are. I will rephrase my change request in hope to clarify the former and explain the latter.

`vector<T>::emplace_back(x)` is essentially `vector<T>::push_back(T(x))` that avoids creation of a temporary T(x) object in certain cases. It is primarily meant to be used in hope to optimize expensive T copies away (via placement new, of all things!), but it can also be used for convenience/readability sake to remove `T`, `(`, and `)` symbols from the function call.

A `vector<T>::emplace_back(T(x))` construct violates the very idea of an emplace API. The reader sees that the code explicitly disables optimizations provided by emplace and purposefully adds symbols made unnecessary by emplace. The reader is left to wonder what is really going on here -- why is this code doing all those things?!

In our specific case, the optimizations are not really important because we are talking about creation of integers -- a temporary on-stack integer costs nearly nothing. If anything, this optimization irrelveance only worsens the reader confusion.

There are many ways to address this problem, including the two snippets I offered in my change request:

```C++
array.emplace_back(intVal);
```

The reader may say: "Ah, OK, the code knows that `array` stores items of a different type than `intVal` and implies that the type conversion (from `intVal` type to the stored type, happening during the hidden stored object construction) is intentional. Good, moving on."

```C++
array.push_back(int(intVal));
```

The reader may say: "Ah, OK, the code knows that `array` stores `int` and tells me that the type conversion (from intVal to `int`, happening right in front of our eyes) is intentional here. Good, moving on."

The first solution is better if you want to write code that is easier to maintain/adjust -- that emplace_back() call will not need to be changed if the type of values stored in the container changes a little (if the type changes in incompatible ways, the compiler will catch the problem).

The second solution is better if you prefer to keep things simple/explicit even if it hurts code maintainability a little.

I am OK with either (or any better) solution in this particular case.

Reply:
Thank you. Not sure how I missed considering that factor.  Fixed in acdf250d2.